### ISSUE TITLE
use cal_ccf and cal_drift with LFC files

### ISSUE THREAD

**Issue by clairem789** [njcuk9999/apero-drs#671]
I have added in allowed DPR types, LFC_LFC for these two recipes in core/instruments/spirou/default_constants.py . 

It seemed it did the trick for the drift recipe; do you think there may be hidden problems?

For the CCF recipe, it gives an error , seemingly due to the BERV calculation (see message below). I'm not sure how to set it up as for the FP_FP, not to consider BERV at all?

08:25:32.204-*|cal_ccf_spirou| Calculating Barycentric Corrections
08:25:32.261- |cal_ccf_spirou| 	 Skipping Barycentric correction (type=LFC_LFC)
08:25:32.644- |cal_ccf_spirou| Wavelength solution read from infile.
08:25:32.645- |cal_ccf_spirou| 	 Wave file = 262101F5T6c_pp_e2dsff_AB_wave_night_AB.fits
08:25:45.467- |cal_ccf_spirou| Reading blaze dymap file:
08:25:45.467- |cal_ccf_spirou| 	/net/GSP/nas12c/big_spirou/FULL_REDUCTION_CALIBDB_06131/science_targets/LFC/calibDB/262100F1T5f_pp_blaze_AB.fits
08:25:45.524-*|cal_ccf_spirou| Computing CCF on fiber AB
08:25:45.581-@|cal_ccf_spirou| W[09-020-00006]: RV not found in header. Setting input RV to 0.0 
08:25:45.582-@|cal_ccf_spirou| 	 Key=OBJRV) 
08:25:45.582-@|cal_ccf_spirou| File =
08:25:45.582-@|cal_ccf_spirou| 	/net/GSP/nas12c/big_spirou/FULL_REDUCTION_CALIBDB_06131/science_targets/LFC/reduced/2021-07-02/2621018a_pp_e2dsff_AB.fits
08:25:45.682-*|cal_ccf_spirou| On fiber AB estimated RV uncertainty on spectrum is 0.017
08:25:46.354- |cal_ccf_spirou| Template used for CCF computation: /
08:25:46.418- |cal_ccf_spirou| Processing CCF fiber AB for Order 0
08:25:46.485-!|cal_ccf_spirou| E[01-010-00001]: Unhandled error has occurred: Error <class 'TypeError'> 
08:25:46.486-!|cal_ccf_spirou| 
08:25:46.486-!|cal_ccf_spirou| Traceback (most recent call last):
08:25:46.487-!|cal_ccf_spirou|  File "/net/GSP/spirou/APERO/apero-drs/apero/core/core/drs_startup.py", line 355, in run
08:25:46.488-!|cal_ccf_spirou|    llmain = func(recipe, params)
08:25:46.488-!|cal_ccf_spirou|  File "/net/GSP/spirou/APERO/apero-drs/bin/cal_ccf_spirou.py", line 200, in __main__
08:25:46.488-!|cal_ccf_spirou|    rv_props1 = velocity.compute_ccf_science(params, recipe, *cargs)
08:25:46.489-!|cal_ccf_spirou|  File "/net/GSP/spirou/APERO/apero-drs/apero/science/velocity/general.py", line 741, in compute_ccf_science
08:25:46.489-!|cal_ccf_spirou|    fit_type, fiber)
08:25:46.489-!|cal_ccf_spirou|  File "/net/GSP/spirou/APERO/apero-drs/apero/science/velocity/general.py", line 1093, in ccf_calculation
08:25:46.490-!|cal_ccf_spirou|    wave_shifts = mp.relativistic_waveshift(rv_ccf - berv)
08:25:46.490-!|cal_ccf_spirou| TypeError: unsupported operand type(s) for -: 'float' and 'NoneType'
08:25:46.491-!|cal_ccf_spirou| 
08:25:46.836-*|cal_ccf_spirou| ***************************************************************************
08:25:46.891-@|cal_ccf_spirou| W[40-003-00005]: Recipe cal_ccf_spirou has NOT been successfully completed
08:25:46.946-*|cal_ccf_spirou| ***************************************************************************

**Comment by njcuk9999** [comment]
## Problem 1: No need to edit the default_constants.py

You should never need to change the github code for changes to constants - unless we are ready to commit something permanent to the code.

For constants this is exactly what your user_constants.ini file is for!
This is located at the DRS_UCONFIG  (do `echo $DRS_UCONFIG` path) 

I'm assuming you changed `CCF_ALLOWED_DPRTYPES` this is the same constant in the user_constants.ini file (if your user_constants.ini is untouched this should just mean uncommenting line 175:
```python
# Allowed input dprtypes for input for ccf recipe
#	dtype=string default=OBJ_DARK, OBJ_FP
# CCF_ALLOWED_DPRTYPES = OBJ_DARK, OBJ_FP
```

So you should revert this change in the `default_constants.py` unless you are going to submit pull requests to change actual code for everyone - we want to avoid at all costs different code in different places its dangerous!

That will at least allow you to force which ever DPRTYPE would want to the ccf recipe... (but that isn't enough for it to work - see problem 2).


## Problem 2: ccf recipe understanding OBJ vs FP vs LFC in different fibers

Currently the recipe only expects "OBJECT" in the AB fiber  and "FP" or "DARK" in the C fiber.

It is in no way clever at all so cannot handle other DPRTYPEs (It currently it wouldn't even be able to handle an FP_FP)  - so we'd have to re-write some code.

Some of the re-writing would be easy but you'd need to have a "path" for every fiber type you wanted to add:
i.e. OBJ, FP or LFC    for both AB and C -- that would at least allow you to have any combination of {OBJ,FP,LFC}_{OBJ,FP,LFC} 

But then we would have to decide the "set up" for LFC which isn't handled on either fiber - would you expect it to be handled exactly as an FP is?

For example:

For `OBJ`:
- we have the BERV
- we "fit" using constant `CCF_FIT_TYPE`  (0 by default = absorption features)
- we use the following constants:
      - CCF_NOISE_SIGDET
      - CCF_NOISE_BOXSIZE
      - CCF_NOISE_THRES
      - CCF_MASK_MIN_WEIGHT
      - CCF_MASK_WIDTH
      - CCF_MASK_UNITS
      - CCF_N_ORD_MAX
- we deal with target rv

For `FP`:
- the BERV is set to 0.0 (see your problem above)
- we "fit" using CCF_FIT_TYPE=1  (emission features)
- we use the following constants:
      - WAVE_CCF_NOISE_SIGDET
      - WAVE_CCF_NOISE_BOXSIZE
      - WAVE_CCF_NOISE_THRES
      - WAVE_CCF_STEP
      - WAVE_CCF_WIDTH
      - WAVE_CCF_TARGET_RV
      - WAVE_CCF_MASK
      - WAVE_CCF_N_ORD_MAX
- we set target rv = 0.0


So we'd have to check all the `FP` settings were fine for `LFC` - or define a new set of CCF values for LFC

But even if all values for FP are fine - in its current form that would only work for the C fiber - I would still need to change code for AB to work with anything other than `OBJ`


### In summary

Don't change the code just to play with a constant - use the user_constants.ini it is exactly what it is for!

The changes you require are all doable it just takes someone to take the time to understand these variables and whether they would work for LFC or how to adjust them for LFC - I'm pretty sure once we established this I could re-code the CCF to allow you to use any combination of `{OBJ,FP,LFC}_{OBJ,FP,LFC}` - the CCF has remained unchanged from 0.6 to 0.7 so we could probably make the changes to both.

**Comment by njcuk9999** [comment]
As for `cal_drift`  this essentially may answer some of our questions - it does the "FP" part of cal_ccf for any fiber - so as long as the settings for FP are fine for LFC the drift would work - however one should still check that all the FP settings are valid for the LFC - especially that you cannot change the mask for the cal_drift recipe - so it will essentially treak the LFC as a FP file

If we were to have a setup in cal_ccf for LFC we could also make cal_drift use this setup.

**Comment by clairem789** [comment]
OK, got it.
-I reverted my changes back in the default_constant file 
-I think that the only change to have cal_drift work for LFC_LFC files (and we may have had FP_LFC and LFC_FP as well, no other combination to my knowledge) would be to allow for another mask. The other constants you mention above are fine with the LFC (emission lines, berv=rv=0 and noise parameters), but the mask is not. I have a mask for the LFC that I can share when we're ready for that development. If the combination of orders for the final RV drift is done with more parameters than the mask itself, we should also adapt this (for instance, the LFC mask does not extend all the way in the K band, so the last orders should not be included when combining CCFs). Also it may be interesting to get several RV values as output (RV in the anchor range, RV out of it, total RV). 
At some point we could plan these developments. Let me know.
Thanks!

**Comment by njcuk9999** [comment]
Yes I think these would be good options - and if it works for cal_drift it should work for cal_ccf too (its the same functionality) - I think it would be great to make both more generalized (i.e. even with just allowing a custom mask in cal_drift you wouldn't be able to use `FP_LFC` or `LFC_FP`)

I think this will have to wait until v0.8(or a future post stable release of 0.7) - but do bug me (after 0.7 is out and being used) for this as it sounds like a good feature to have!
